### PR TITLE
HDDS-9330. Optimize container lookup by owner during block allocation

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

Allocate ratis block latency optimization

When applying for blocks, if the number of containers under a pipeline is high, it will consume a lot of time in filtering the owner's part.

This pr optimizes the time consumed by filtering owners, and after testing in real scenarios.

See jira for performance details.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9330

## How was this patch tested?

existing UT

ci see: https://github.com/guohao-rosicky/ozone/actions/runs/6248198048

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
cc @ChenSammi PTAL.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java [hunk 1]
```diff
@@ -362,18 +362,10 @@ private NavigableSet<ContainerID> getContainersForOwner(
...
-      try {
-        if (!getContainer(cid).getOwner().equals(owner)) {
-          containerIDIterator.remove();
-        }
-      } catch (ContainerNotFoundException e) {
-        LOG.error("Could not find container info for container {}", cid, e);
-        containerIDIterator.remove();
-      }
-    }
+    containerIDs = new TreeSet<>(containerIDs);
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:365]: Why are you creating a `new TreeSet<>(containerIDs)` here? The call to `pipelineManager.getContainersInPipeline()` already returns the new TreeSet.

**Comment by sodonnel** [comment]
Based on the charts in the Jira, this seems like a good change. I am surprised it is so much faster.

You said you have about 200 containers per pipeline. Assuming to total number of containers for an owner in the system is much larger than that, lets say 2000. You need to form a new set of 2000 elements, and then do 200 lookups into that set to run the `retainAll()` method.

The current code iterates the 200 and then:

1. Looks up the containerInfo in a map - I'd guess at a similar expense to the 200 probes into the map for retain all.
2. Then for each containerInfo it performs the equals comparison against owner.

It would therefore suggest that its the `equals()` calls that are expensive and much more so than the lookups into the map.

In the tests you were running, there are 200 containers per pipeline:

1. How many owners are there?
2. How many open pipelines are there in the cluster?
3. Do you know the total distribution of pipelines per owner?

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java (hunk 1 above)
Comment (guohao-rosicky) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java:365]: > Why are you creating a `new TreeSet<>(containerIDs)` here? The call to `pipelineManager.getContainersInPipeline()` already returns the new TreeSet.

Sorry, it really should be removed, I didn't notice it

**Comment by guohao-rosicky** [comment]
Hi @sodonnel Thank you for your comment, I describe, my test environment, and how I localized the issue

> In the tests you were running, there are 200 containers per pipeline:
> 
> 1. How many owners are there? & 2. How many open pipelines are there in the cluster?
> 3. Do you know the total distribution of pipelines per owner?


1. This is a test cluster with 4 data nodes and 36 disks for each data node
2. there are about dozens of owners, this is due to my version, the use of owners of multi-tenant division, the actual community for the current version will not have a lot of owners
3. pipeline 44

**Comment by guohao-rosicky** [comment]
@sodonnel  Here's how I'm situating this

1. I am using https://github.com/alibaba/arthas to trace the allocate block
2. I found that the main time consuming part is at https://github.com/apache/ozone/blob/master/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ container/ContainerManagerImpl.java#L369

getContainer method, about 0.5ms per call through arthas, if you need to iterate 200 times, it will take more time.



Translated with www.DeepL.com/Translator (free version)

**Comment by guohao-rosicky** [comment]
hi, @sodonnel Thanks for the continued focus on performance.

> I wonder how this would work if you had only 2 owners, so the set has to be formed from 4400 containers. Perhaps it will still be faster than the containerInfo lookup and owner check the current code performs.

According to the original logic, it doesn't seem to matter how many owners there are.
Because it will iterate through all of them and return

```
private NavigableSet<ContainerID> getContainersForOwner(
      Pipeline pipeline, String owner) throws IOException {
    NavigableSet<ContainerID> containerIDs =
        pipelineManager.getContainersInPipeline(pipeline.getId());
    Iterator<ContainerID> containerIDIterator = containerIDs.iterator();
    while (containerIDIterator.hasNext()) {
      ContainerID cid = containerIDIterator.next();
      try {
        if (!getContainer(cid).getOwner().equals(owner)) {
          containerIDIterator.remove();
        }
      } catch (ContainerNotFoundException e) {
        LOG.error("Could not find container info for container {}", cid, e);
        containerIDIterator.remove();
      }
    }
    return containerIDs;
  }
```

**Comment by adoroszlai** [comment]
@guohao-rosicky please fix checkstyle failure

```
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java
 30: Unused import - java.util.TreeSet.
```

https://github.com/guohao-rosicky/ozone/actions/runs/6269859519/job/17026915840

**Review by adoroszlai — COMMENTED** [review]
Thanks @guohao-rosicky for the improvement, LGTM.

I'd change the title of the PR to `HDDS-9330. Optimize container lookup by owner during block allocation`, because I think the original one is too generic.  (I guess this is not the only improvement for block allocation latency.)  Let me know what you think.

**Comment by guohao-rosicky** [comment]
> HDDS-9330. Optimize container lookup by owner during block allocation

Hi, @adoroszlai, Thank you for your attention. This title, explains this PR very well, I will change it.

**Review by adoroszlai — DISMISSED** [review]

**Comment by adoroszlai** [comment]
@sodonnel would you like to take another look?

**Comment by sodonnel** [comment]
@adoroszlai I am happy for this to be committed. It will probably make things better than they are currently, but it still feels like a very expensive way to get the containers on a pipeline for a given owner. The new code will be forming a set of all containers for an owner across the whole cluster, and using that to filter the owners for a single pipeline.

I also wonder about the usual case where a cluster has only 1 owner. Does this code path still get followed? I have not had a chance to check yet.

**Comment by adoroszlai** [comment]
> it still feels like a very expensive way to get the containers on a pipeline for a given owner. The new code will be forming a set of all containers for an owner across the whole cluster, and using that to filter the owners for a single pipeline.

I was wondering the same.  But "containers by owner" map already exists:

https://github.com/apache/ozone/blob/f5d9223f99f1f4fbc76ba3c979dbceafd5fd83b6/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerStateMap.java#L88

we're just making an immutable copy of the set stored as the value in the map:

https://github.com/apache/ozone/blob/f5d9223f99f1f4fbc76ba3c979dbceafd5fd83b6/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerAttribute.java#L192-L193

Would it be better keep the original iteration and use `hasContainerID` instead?

https://github.com/apache/ozone/blob/f5d9223f99f1f4fbc76ba3c979dbceafd5fd83b6/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerAttribute.java#L113-L126

**Comment by sodonnel** [comment]
Looking at where this code is called, we call it from WriteableRatisContainerProvider, where it calls:

```
    containerInfo = containerManager.getMatchingContainer(size, owner,
        pipeline, excludeList.getContainerIds());
```
Where it may call getContainersForOwner twice under a synchronised block:

```
  @Override
  public ContainerInfo getMatchingContainer(final long size, final String owner,
      final Pipeline pipeline, final Set<ContainerID> excludedContainerIDs) {
    NavigableSet<ContainerID> containerIDs;
    ContainerInfo containerInfo;
    try {
      synchronized (pipeline.getId()) {
        containerIDs = getContainersForOwner(pipeline, owner);
        if (containerIDs.size() < getOpenContainerCountPerPipeline(pipeline)) {
          allocateContainer(pipeline, owner);
          containerIDs = getContainersForOwner(pipeline, owner);
        }
        containerIDs.removeAll(excludedContainerIDs);
        containerInfo = containerStateManager.getMatchingContainer(
            size, owner, pipeline.getId(), containerIDs);
        if (containerInfo == null) {
          containerInfo = allocateContainer(pipeline, owner);
        }
        return containerInfo;
      }
    } catch (Exception e) {
      LOG.warn("Container allocation failed on pipeline={}", pipeline, e);
      return null;
    }
  }
```

In the "normal" case, as pipeline will have relatively few containers and most cluster I have seen only have a single owner. The cluster described here has about 200 containers per pipeline and dozens of owners, so it is very far away from usual.

Consider a large cluster with several 1000 open containers, and with many open pipelines, so that there are only a small number of containers pipeline, lets guess 10.

In the old code path, we would have to iterate 10 times and check the 10 containers, but in this new code path, we will form a map of all containers in the cluster (they all have the same owner), to filter nothing and return.

So my concern is that this change will make the normal code path slower to make this unusual path faster. As this code is part of the container selection process on write, anything that slows it down is going to be bad.

In fact, when you look at this "usual case" where a cluster only has a single owner for everything, even the existing code path is doing a lot of work it does not need to do.

It would be worth trying the suggestion from @adoroszlai to probe the existing containersByOwner map to see if it performs better than looking up the containerInfo and doing a compare on the owner as it does currently.

I was never aware of how the "owner" concept was supposed to be used, but in Cloudera we have not had any use cases to use it, and it looks like we pay a performance price anyway. Perhaps we need to have a way to turn this off to be more efficient if multiple owners are not in use on a cluster.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManagerImpl.java [hunk 2]
```diff
@@ -290,6 +290,16 @@ public Set<ContainerID> getContainerIDs(final LifeCycleState state) {
...
   }
 
+  @Override
+  public Set<ContainerID> getContainerIDs(final String owner) {
+    lock.readLock().lock();
+    try {
+      return containers.getContainerIDsByOwner(owner);
+    } finally {
+      lock.readLock().unlock();
+    }
```
Comment (szetszwo) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManagerImpl.java:300]: Use the `readLock()` method.
```java
    try (AutoCloseableLock ignored = readLock()) {
      return containers.getContainerIDsByOwner(owner);
    }
```

**Review by szetszwo — COMMENTED** [review]
@guohao-rosicky , this is really a good idea!

We may also improve the `getCollection` method to reduce one lookup.
```java
+++ b/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerAttribute.java
@@ -189,8 +189,9 @@ public boolean remove(T key, ContainerID value) {
   public NavigableSet<ContainerID> getCollection(T key) {
     Preconditions.checkNotNull(key);
 
-    if (this.attributeMap.containsKey(key)) {
-      return ImmutableSortedSet.copyOf(this.attributeMap.get(key));
+    final NavigableSet<ContainerID> set = attributeMap.get(key);
+    if (set != null) {
+      return ImmutableSortedSet.copyOf(set);
     }
     if (LOG.isDebugEnabled()) {
```

**Comment by szetszwo** [comment]
@guohao-rosicky , In order to address @sodonnel 's concern, let's use a merge-sort like algorithm to compute the intersection instead of using `retainAll`; see https://issues.apache.org/jira/secure/attachment/13063614/5335_idea.patch

**Comment by guohao-rosicky** [comment]
> @guohao-rosicky , In order to address @sodonnel 's concern, let's use a merge-sort like algorithm to compute the intersection instead of using `retainAll`; see https://issues.apache.org/jira/secure/attachment/13063614/5335_idea.patch

Thanks @szetszwo  for looking, that's a good idea, I'll apply the patch.

**Comment by adoroszlai** [comment]
/pending that's a good idea, I'll apply the patch

**Pull request closed without merging**